Gabows connectivity inspector (followup)

jkinable opened the pull request:
This is a followup of the old 2013 pull request number 53: https://github.com/jgrapht/jgrapht/pull/53

Although all the code was provided, the merge into jgrapht has never been completed due to a rebasing issue. I reworked the code and made some minor modifications.
-Added GabowStrongConnectivityInspector (alg), StrongConnectivityInspector (Interface), GabowStrongConnectivityTest (test class).

There used to be an algorithm class called StrongConnectivityInspector.java This name is however very confusing, so for consistency I renamed it to KosarajuStrongConnectivityInspector. So now we have:
StrongConnectivityInspector (interface)
-GabowStrongConnectivityInspector (implementation of StrongConnectivityInspector)
-KosarajuStrongConnectivityInspector (implementation of StrongConnectivityInspector)

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/GabowStrongConnectivityInspectorTest.java:
```diff
@@ -0,0 +1,285 @@
+ * the Eclipse Foundation.
+ */
+/* ------------------------------
+ * ConnectivityInspectorTest.java
```
This section was copy/pasted incorrectly.

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/GabowStrongConnectivityInspectorTest.java:
```diff
@@ -0,0 +1,285 @@
+
+/**
+ * Test cases for the GabowStrongConnectivityInspector. Tests are identical to the tests for the KosarajuStrongConnectivityInspector as provided in the
+ * ConnectivityInspectorTest class.
```
Instead of copy/pasting the entire test, refactor the existing test so that it can be reused across different algo implementations.

jkinable replied:
Do you have a recommendation on how to do this? I agree that this would be cleaner, but looking at the test cases there are quite a lot of different instances being created of the various StrongConnectivityInspectors. I can't just pass a connectivityInspector as a parameter because you need to specify the graph in the constructor. So I don't really know how to do this.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/GabowStrongConnectivityInspector.java:
```diff
@@ -0,0 +1,260 @@
+package org.jgrapht.alg;
```
All files need the standard JGraphT header (including both the project-level header and the file-level header).

jsichi commented:
Per project policy, breaking changes (such as changing StrongConnectivityInspector from an implementation class to an interface) should be avoided.  Instead, we always deprecate in release X, and then it's OK to finally remove in release X+1.  (This makes the upgrade path for dependent projects much smoother.)  So in this case, it's fine to move the implementation logic to KosarajuStrongConnectivityInspector, but:

* the StrongConnectivityInspector class needs to stay around for one release (marked as deprecated)

* consequently, the new interface should have a different name to avoid confusion (StrongConnectivityAlgorithm?)

jkinable commented:
I've processed all your requests. I've removed the StrongConnectivity tests from the ConnectivityInspectorTest.java class and moved them to a new test class (StrongConnectivityAlgorithmTest.java). This new test class now tests both the GabowStrongConnectivityInspector and the KosarajuStrongConnectivityInspector.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/StrongConnectivityInspector.java:
```diff
@@ -58,9 +60,10 @@
  * @author Christian Hammer
  * @since Feb 2, 2005
```
Add a @deprecated tag to the Javadoc as well, instructing the developer to use Kosaraju instead.

jsichi commented:
The file header is still missing on GabowStrongConnectivityInspector.java.

Other than that (and adding deprecated in the Javadoc as well), it's good to go.

jkinable commented:
Not sure how I missed that one. It should be there now :)
